### ISSUE TITLE
[TF2.0] Can't use keras validation data with distribution strategy 

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#28797]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): pip tf-nightly-gpu-2.0-preview
- TensorFlow version (use command below): 2.0.0.dev20190517
- Python version: 3.7.3
- CUDA/cuDNN version: 10 / 7.4.2.24
- GPU model and memory: 4 x V100

**Describe the current behavior**
Using keras validation data and `tf.distribute.MirroredStrategy()` will fail with `'BatchDataset' object is not subscriptable` in TF 2.0 nightly.

**Describe the expected behavior**
Without distribution strategy everything works fine.

**Code to reproduce the issue**
```python
import tensorflow as tf
import tensorflow_datasets as tfds

mnist_train, mnist_test = tfds.load(name='mnist', split=[tfds.Split.TRAIN, tfds.Split.TEST], as_supervised=True)

strategy = tf.distribute.MirroredStrategy()

def scale(image, label):
  image = tf.cast(image, tf.float32)
  image /= 255

  return image, label

train_dataset = mnist_train.map(scale).shuffle(1000).batch(256)
test_dataset = mnist_test.map(scale).batch(256)

with strategy.scope():
  model = tf.keras.Sequential([
      tf.keras.layers.Conv2D(32, 3, activation='relu', input_shape=(28, 28, 1)),
      tf.keras.layers.MaxPooling2D(),
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(64, activation='relu'),
      tf.keras.layers.Dense(10, activation='softmax')
  ])

  model.compile(loss='sparse_categorical_crossentropy',
                optimizer=tf.keras.optimizers.Adam(),
                metrics=['accuracy'])

model.fit(train_dataset, validation_data=test_dataset, epochs=10)
```

**Other info / logs**
```python traceback
  File "tf_bug.py", line 30, in <module>
    model.fit(train_dataset, validation_data=test_dataset, epochs=10)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training.py", line 649, in fit
    validation_freq=validation_freq)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training_distributed.py", line 143, in fit_distributed
    steps_name='steps_per_epoch')
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training_arrays.py", line 145, in model_iteration
    _print_train_info(inputs, val_inputs, steps_per_epoch, verbose)
  File "/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/engine/training_arrays.py", line 450, in _print_train_info
    hasattr(inputs[0], 'shape') and hasattr(val_inputs[0], 'shape')):
TypeError: 'BatchDataset' object is not subscriptable
```

**Comment by muddham** [comment]
@lgeiger Able to reproduce the issue with the provided code.

TypeError: 'BatchDataset' object does not support indexing

**Comment by jvishnuvardhan** [comment]
@lgeiger I don't see any issue when I ran the code with TF2.0 (gist is [here](https://colab.sandbox.google.com/gist/jvishnuvardhan/aca319db01cf71284a0dc7b8001be7f7/untitled191.ipynb) as well as TF-nightly-gpu-2.0-preview(gist is [here](https://colab.sandbox.google.com/gist/jvishnuvardhan/6d1c69fedc2846663a94d388bd0796ce/untitled192.ipynb). Thanks!

**Comment by omalleyt12** [comment]
I'm also unable to repro, @lgeiger could you try the lastest nightly? This may be already fixed

**Comment by lahertel** [comment]
@omalleyt12 I am still able to reproduce the error even with the latest nightly (2.0.0.dev20190523)

**Comment by lgeiger** [comment]
@jvishnuvardhan @omalleyt12 Have run this code on multiple GPUs?

**Comment by lahertel** [comment]
Updating to the latest nightly (2.0.0.dev20190528), I receive a different error message:
```python
  File "train_multi.py", line 60, in <module>
    verbose=verbose)
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 644, in fit
    use_multiprocessing=use_multiprocessing)
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_distributed.py", line 899, in fit
    validation_freq=validation_freq)
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_distributed.py", line 149, in fit_distributed
    steps_name='steps_per_epoch')
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_arrays.py", line 415, in model_iteration
    steps_name='validation_steps')
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_arrays.py", line 178, in model_iteration
    steps_per_epoch)
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_arrays.py", line 469, in _get_num_samples_or_steps
    'steps_per_epoch')
  File "~/.virtualenvs/tf_v2_nightly/lib/python3.7/site-packages/tensorflow/python/keras/engine/training_utils.py", line 256, in check_num_samples
    if hasattr(ins[0], 'shape'):
TypeError: 'function' object is not subscriptable
```

**Comment by omalleyt12** [comment]
I was able to reproduce this new error, looking into a fix now. In the meantime a workaround should be to pass the `validation_steps` arg to `fit`

**Comment by guptapriya** [comment]
Similar error here: https://github.com/tensorflow/tensorflow/issues/30843#issuecomment-514534349

**Comment by omalleyt12** [comment]
This is fixed in the latest 2.0 nightly
